connectivity: Add IPsec key derivation validation integration test

pillai-ashwin opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer's Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

This PR adds integration test coverage for IPsec key derivation validation across multiple nodes. The new test ensures that Cilium's IPsec key derivation algorithm produces consistent cryptographic keys for the same tunnels on different nodes.

**What this PR does:**
- Adds `IPsecKeyDerivationValidation` connectivity test that extracts XFRM states from multiple Cilium nodes
- Validates that derived keys (AEAD, authentication, encryption) are identical across nodes for matching tunnels
- Covers both IPv4 and IPv6 tunnels (node-to-node and pod-to-pod)
- Integrates seamlessly into existing IPsec test infrastructure without creating separate files
- Runs as part of concurrent connectivity tests when IPsec encryption is enabled

**Testing:**
- Validated in kind cluster with IPsec enabled
- Successfully validates 8 tunnel key pairs across multiple address ranges
- Provides detailed logging for debugging key derivation issues

Fixes: #35843

```release-note
Add integration test to validate IPsec key derivation consistency across nodes, ensuring tunnel keys are properly derived and identical between Cilium pods.
```

pchaigno commented on cilium-cli/connectivity/builder/builder.go:
```diff
@@ -278,6 +278,7 @@ func concurrentTests(connTests []*check.ConnectivityTest) error {
 		podToPodEncryption{},
```
*Commenting on the commit description itself*

Please copy the context from the pull request description to the commit description. See https://pchaigno.github.io/cilium/2024/12/05/first-cilium-pull-request.html#preparing-the-branch for what I'd generally recommend and an example.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+
+	ciliumPods := ct.CiliumPods()
+	if len(ciliumPods) < 2 {
+		testContext.Logf("Not enough cilium pods for multi-node key derivation test (have %d, need 2)", len(ciliumPods))
```
I think we can error here as the connectivity tests probably can't start without two Cilium agents (so this should never happen).

pillai-ashwin replied:
Done!

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+// getXfrmStates extracts XFRM state information from a cilium pod
+func (t *ipsecKeyDerivationTest) getXfrmStates(ctx context.Context, testContext *check.Test, pod check.Pod) ([]xfrmStateInfo, error) {
+	// Use the ip command to extract XFRM states with detailed key information
+	cmd := []string{"ip", "xfrm", "state", "list"}
```
Why doesn't `netlink.XfrmStateList(netlink.FAMILY_ALL)` work?

pillai-ashwin replied:
did another implementation with two parts - 
1. running cilium-dbg encrypt states command (inside pod)
2. remote command execution (from test)

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+	if len(nodeIPs) < 2 {
+		testContext.Logf("Not enough nodes with XFRM states for validation (have %d, need 2)", len(nodeIPs))
+		return
+	}
```
Not sure this check is needed given we already checked the number of Cilium agents, but if we're keeping it, we can move it before we create `nodeIPs`.

pillai-ashwin replied:
I've removed the redundant check entirely. 

Since we already verify `len(ciliumPods) >= 2` at the start of `Run()`, and we iterate through all those pods to build `nodeStates`, the second check was indeed unnecessary. If any pod fails to return XFRM states, we'll get a clear error from the `getXfrmStates()` call anyway.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+			// Find matching tunnels between the two nodes
+			matches := t.findMatchingTunnels(node1States, node2States, node1IP, node2IP)
+			if len(matches) == 0 {
+				testContext.Logf("No matching tunnels found between nodes %s and %s", node1IP, node2IP)
```
I think we should fail the test when this happens. All Cilium-managed node pairs should have matching XFRM states.

pillai-ashwin replied:
failing the test when no matching tunnels are found

```go
if matchingPairs == 0 {
    testContext.Fatalf("No matching tunnels found between nodes %s and %s - all Cilium-managed node pairs should have matching XFRM states", node1IP, node2IP)
}
```

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+// statesRepresentSameTunnel checks if two XFRM states represent the same tunnel
+func (t *ipsecKeyDerivationTest) statesRepresentSameTunnel(state1, state2 xfrmStateInfo) bool {
+	// States represent the same tunnel if they have the same src, dst, and direction
+	return state1.Src == state2.Src && state1.Dst == state2.Dst
```
Shouldn't we check the directions are opposites as well?

pillai-ashwin replied:
Made some changes. Looking at the current implementation, we're actually doing this correctly for key derivation validation. The current logic:

```go
// Index node2 states by src->dst for efficient lookup
node2Index := make(map[string]xfrmStateInfo)
for _, state := range node2States {
    if state.ReqID == 1 { // Only Cilium-managed states
        key := state.Src + "->" + state.Dst
        node2Index[key] = state
    }
}

// Find matching tunnel states using index lookup
for _, state1 := range node1States {
    if state1.ReqID == 1 {
        key := state1.Src + "->" + state1.Dst
        if state2, exists := node2Index[key]; exists {
            // Compare keys for same tunnel direction
        }
    }
}
```

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+	var matches []tunnelMatch
+
+	for _, state1 := range states1 {
+		for _, state2 := range states2 {
```
Could we index the states for each node by their `Src` and `Dst`? That way we'd avoid this second `for` loop.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+				})
+			}
+		}
+	}
```
I think we should also fail if we find any state in one node that doesn't have a match in the other.

pillai-ashwin replied:
Nice! That's a great performance optimization. I've implemented the indexing approach to avoid the nested loop

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,263 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+			testContext.Fatalf("Failed to get XFRM states from node %s: %v", hostIP, err)
+		}
+		nodeStates[hostIP] = states
+		testContext.Logf("Found %d XFRM states on node %s", len(states), hostIP)
```
I'd like to allow users to run this test on large clusters. Which means we probably want to switch all log messages that are emitted per-node to be debug messages (`.Debugf()`).

pillai-ashwin replied:
Done!

pchaigno reviewed (commented):
This is fantastic! Thanks a lot for working on this!

I've left comments below. I think right now the two blockers are:
- I'd much prefer if we can avoid parsing text output and instead rely on the `netlink` library as usual.
- I think this would be particularly useful as a debugging tool, which means we may need to run it on large clusters. So we need to rethink iterations and logging a bit (see below).

derailed commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -107,6 +126,69 @@ func dumpIPsecStatus() (models.EncryptionStatus, error) {
 	return status, nil
 }
 
+type XfrmStateInfo struct {
```
Looks like this is duplicating `xfrmStateJSON` struct above. Couldn't we reuse it here?

pillai-ashwin replied:
did that to avoid import cycles between `cilium-dbg/cmd` and `cilium-cli` packages through their shared dependency on `pkg/datapath/loader`

I've renamed it to `xfrmStateInfo` and added a comment there stating why I defined it locally.

pchaigno replied:
Could we move this to `pkg/common/ipsec/utils.go` and import it from both places?

pillai-ashwin replied:
done

derailed commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,239 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+	// The keys should be identical for states representing the same tunnel
+	// This validates that the key derivation algorithm produces consistent results
+
+	if state1.AuthKey != "" && state2.AuthKey != "" {
```
Perhaps we could move this out and just add a method on xfrmStateInfo to validate the keys?

pillai-ashwin replied:
Nice! I've refactored the validation logic.

derailed reviewed (commented):
@pillai-ashwin Nice work! a bit out of context here so feel free to discard...

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -4,6 +4,7 @@
 package cmd
```
Could you move the introduction of the new `cilium-dbg` command into its own first commit, explaining why we'll need this in the second commit?

pillai-ashwin replied:
done

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -51,9 +52,27 @@ var encryptStatusCmd = &cobra.Command{
 }
 
+var encryptStatesCmd = &cobra.Command{
+	Use:   "states",
```
I understand we need some way to dump states from within the Cilium pods, but that command will be confusing given the existing `status` one.

I don't expect users to ever need to use this. Most users shouldn't dig that low into implementation details and advanced users can just use `ip xfrm state`. An alternative to move this away from any user interaction would be to make it an agent API, but then we'd have a dependency on the agent being up.

So let's keep it as is, but rename to something like `dump_xfrm` and clarify in the description that it only exists to facilitate the collection of structured XFRM states as part of tests.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+
+			testContext.Debugf("Validating key derivation between nodes %s and %s", node1IP, node2IP)
+
+			// Index node2 states by src->dst for efficient lookup (as requested by pchaigno)
```
No need to mention folks in code comments :) It doesn't help understand the code.

pillai-ashwin replied:
oops, haha - missed that one :)

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+			// Index node2 states by src->dst for efficient lookup (as requested by pchaigno)
+			node2Index := make(map[string]xfrmStateInfo)
+			for _, state := range node2States {
+				if state.ReqID == 1 { // Only Cilium-managed states
```
The `cilium-dbg` command already filters those out, so can we drop all such checks from the cilium-cli command?

pillai-ashwin replied:
done

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+			}
+
+			// Find and validate matching tunnel states using index lookup
+			matchingPairs := 0
```
This can be a boolean.

pillai-ashwin replied:
done

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+type xfrmStateInfo struct {
+	Src      string `json:"src"`
+	Dst      string `json:"dst"`
+	SPI      uint32 `json:"spi"`
```
We collect the SPI here but never use it. We definitely should be using it when checking two states match.

The only time when we can have multiple SPIs is during key rotation. At that time we may have twice as many matching states, with the old SPI, the new one, or both. We should just check that we have at least one matching state for a given SPI.

So I guess we probably want to index states by their SPI in addition to src and dst.

pillai-ashwin replied:
Nice! Added SPI to state matching key for proper key rotation handling

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+					key := state.Src + "->" + state.Dst
+					node2Index[key] = state
+				}
+			}
```
We're going to recompute this for every single node pair. Let's move it out of the second for loop and into the first one, to compute only once per node.

pillai-ashwin replied:
Nice! moved index computation to first loop to avoid recompute

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+			direction := state.Src + "->" + state.Dst
+			tunnelDirections[direction] = true
+		}
+	}
```
Can't we reuse `node2Index` instead of creating `tunnelDirections`?

pillai-ashwin replied:
yeah, reused existing nodeIndex instead of creating new map

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +129,203 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+	t.validateNodeBidirectionalTunnels(testContext, node1States, node1IP)
+
+	// Check bidirectional tunnels on node2
+	t.validateNodeBidirectionalTunnels(testContext, node2States, node2IP)
```
If we perform this check in the first for loop of `validateKeyDerivation` (line 236), then I don't think we'll need to check both `node1States` and `node2States` as we'll already iterate on all nodes.

pillai-ashwin replied:
consolidated the bidirectional validation into main loop

pchaigno reviewed (commented):
Thanks for the new version!

Lots of comments below, but I think we're getting closer and we'll have a very good test in the end.

pchaigno commented on cilium-cli/connectivity/builder/no_ipsec_xfrm_errors.go:
```diff
@@ -17,3 +17,12 @@ func (t noIpsecXfrmErrors) build(ct *check.ConnectivityTest, _ map[string]string
 		WithScenarios(tests.NoIPsecXfrmErrors(ct.Params().ExpectedXFRMErrors))
 }
+
+type ipsecKeyDerivation struct{}
```
First commit: All cilium-cli/ changes should be in the second commit.

pillai-ashwin replied:
done

pchaigno reviewed (commented):
Thanks for the new version! It looks a lot nicer.

I think there's an issue on the commit history right now, so a bit hard to review. I'll do a new review once that's fixed, but hopefully we should be close to merge :)

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -107,6 +129,69 @@ func dumpIPsecStatus() (models.EncryptionStatus, error) {
 	return status, nil
 }
 
+type XfrmStateInfo struct {
```
Can't we include `pkg/common/ipsec` instead of defining this again?

pillai-ashwin replied:
removed duplicate struct using`pkg/common/ipsec`

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -107,6 +129,69 @@ func dumpIPsecStatus() (models.EncryptionStatus, error) {
+	AeadKey  string `json:"aead_key,omitempty"`
+}
+
+func dumpXfrmStates() ([]XfrmStateInfo, error) {
```
This is never used, is it?

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -51,9 +52,30 @@ var encryptStatusCmd = &cobra.Command{
+var encryptDumpXfrmCmd = &cobra.Command{
+	Use:   "dump-xfrm",
+	Short: "Dump structured XFRM states for test facilitation (internal use only)",
+	Long: `Dump structured XFRM states for test facilitation.
```
See the CI failure. You'll need to run `make -C Documentation update-cmdref` to update the cmdref documentation for this new command.

pillai-ashwin replied:
done

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -129,3 +130,193 @@ func (n *noIPsecXfrmErrors) loadIPsecXfrmErrors(t *check.Test) map[string]string
+// xfrmStateInfo represents the key information from an XFRM state
+// This matches the XfrmStateInfo struct from pkg/common/ipsec/utils.go
+// We define it locally here to avoid import cycles
+type xfrmStateInfo struct {
```
Is this still needed? Can't we import `pkg/common/ipsec`?

pchaigno reviewed (commented):
A few small mistakes around the use of `pkg/common/ipsec`.

pchaigno commented on cilium-dbg/cmd/encrypt_status.go:
```diff
@@ -79,7 +100,7 @@ func dumpIPsecStatus() (models.EncryptionStatus, error) {
 		Ipsec: &models.IPsecStatus{},
 	}
-	xfrmStates, err := safenetlink.XfrmStateList(netlink.FAMILY_ALL)
+	xfrmStates, err := safenetlink.XfrmStateList(0) // Use 0 instead of netlink.FAMILY_ALL
```
> Use 0 instead of netlink.FAMILY_ALL

Why?

pillai-ashwin replied:
both `netlink.FAMILY_ALL` and `0` tell netlink to list XFRM states for all address families (IPv4, IPv6, etc. but you're right - `netlink.FAMILY_ALL` is much more readable and self-documenting than 0. I'll revert it.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -9,16 +9,16 @@ import (
-const (
-	maxExpectedErrors = 10
-)
+const maxExpectedErrors = 10
```
Let's keep this as is please.

pillai-ashwin replied:
was thinking i'll simplify the const declaration since there's only one constant but yeah it's unnecessary - I'll keep it as is.

pchaigno commented on cilium-cli/connectivity/tests/ipsec_xfrm.go:
```diff
@@ -52,7 +52,6 @@ func (n *noIPsecXfrmErrors) Run(ctx context.Context, t *check.Test) {
 
 	if ct.Params().ConnDisruptTestSetup {
 		n.storeIPsecXfrmErrors(t, crtXfrmErrors)
-		return
```
Why this change?

pillai-ashwin replied:
Thank you for this catch! 🙏 

I guess I was thinking - why not run the validation in both phases? but after I looked into it realized that in setup phase - there's no previous baseline to compare against...

so without the return i'm guessing the setup phase would try to load a non-existent baseline file and crash on -
`prevXfrmErrors := n.loadIPsecXfrmErrors(t)`

reverted it in the conn-disrupt test.

pchaigno replied:
In general, it's best to not introduced unrelated changes in a commit. If you think something can be improved, change it in a separate commit with the rationale why.

pchaigno reviewed (commented):
There are a couple changes that don't look related to this pull request. Not sure if it's new in last push or if I missed it before.

pchaigno commented on pkg/common/ipsec/utils.go:
```diff
@@ -4,14 +4,83 @@
+}
+
+// DumpXfrmStates extracts XFRM state information using netlink
+func DumpXfrmStates() ([]XfrmStateInfo, error) {
```
There's no reason to expose this to other packages as it's only used in `cilium-dbg/cmd/encrypt_status.go`. Let's move it there.

pillai-ashwin replied:
Done!

pchaigno commented on pkg/common/ipsec/utils.go:
```diff
@@ -4,14 +4,83 @@
+	CryptKey string `json:"crypt_key,omitempty"`
+	AeadAlg  string `json:"aead_alg,omitempty"`
+	AeadKey  string `json:"aead_key,omitempty"`
+}
```
Hm, putting this struct here isn't actually going to work. CI is failing because of this. The Cilium CLI is compiled for Darwin, this file is pulled, and since it contains Linux specific logic (accesses to `XfrmState` fields), it fails to compile. Instead, I think we could move this struct to a new `types.go` file in the same package.

pillai-ashwin replied:
Done!

pchaigno reviewed (changes requested):
The following comments should address the CI failure.

pchaigno commented on pkg/common/ipsec/utils.go:
```diff
@@ -4,14 +4,83 @@
+}
+
+// DumpXfrmStates extracts XFRM state information using netlink
+func DumpXfrmStates() ([]XfrmStateInfo, error) {
```
Why is this introduced in the first commit just to be removed in the second? Please spend more time checking your own changes.

pillai-ashwin replied:
Fixed. Reorganized into 3 logical commits - no code is introduced then removed. The function now only exists in cilium-dbg/cmd/encrypt_status.go as requested.

pchaigno commented on pkg/types/xfrm.go:
```diff
@@ -0,0 +1,20 @@
+	CryptKey string `json:"crypt_key,omitempty"`
+	AeadAlg  string `json:"aead_alg,omitempty"`
+	AeadKey  string `json:"aead_key,omitempty"`
+}
```
As said previously, this change should probably be introduced in its own commit, with an explanation why we needed to extract it into the pkg/types/ package.

pillai-ashwin replied:
Fixed. Moved to separate commit with explanation. This was needed to avoid cross-compilation issues since cilium-cli builds for Darwin (maybe Windows too) but XFRM code contains Linux-specific dependencies.

pillai-ashwin committed:
pkg/types: Add XfrmStateInfo struct for cross-platform compatibility

Extract XfrmStateInfo struct to pkg/types to resolve cross-compilation issues.
This allows cilium-cli (which builds for Darwin/Windows) to import the struct
without pulling in Linux-specific netlink dependencies from cilium-dbg.

Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

pillai-ashwin committed:
cilium-dbg: Add encrypt dump-xfrm command for test facilitation

Add dump-xfrm subcommand to extract XFRM state information in JSON format.
This command is used by integration tests to validate IPsec key derivation
consistency across nodes.

Key features:
- Linux-specific build constraints to prevent cross-compilation issues
- Filters for Cilium-managed states (ReqID == 1)
- Extracts authentication, encryption, and AEAD key information
- JSON output for programmatic consumption by tests

Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

pillai-ashwin committed:
cilium-cli: Add IPsec key derivation validation test

Add integration test to validate IPsec key derivation consistency across nodes.
The test ensures that Cilium's IPsec key derivation algorithm produces identical
cryptographic keys for the same tunnels on different nodes.

Key features:
- Extracts XFRM states from all Cilium nodes using cilium encrypt dump-xfrm
- Validates key consistency (AEAD, authentication, encryption) across nodes
- Checks bidirectional tunnel establishment for all tunnel directions
- Runs as part of concurrent connectivity tests when IPsec is enabled
- Uses SPI-based tunnel correlation for accurate state matching
- Requires Cilium 1.19.0-pre.0+ for the dump-xfrm command

The test helps detect key derivation inconsistencies that could cause
IPsec tunnel failures or security issues in multi-node deployments.

Fixes: #35843
Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

pchaigno commented on cilium-cli/connectivity/tests/ipsec_key_derivation.go:
```diff
@@ -0,0 +1,187 @@
+		}
+
+		if len(states) == 0 {
+			test.Debugf("No XFRM states found on node %s", nodeName)
```
Let's fail if this happens.

pillai-ashwin replied:
Changed to `test.Fatalf("No XFRM states found on node %s", nodeName)`

pchaigno commented on cilium-cli/connectivity/tests/ipsec_key_derivation.go:
```diff
@@ -0,0 +1,187 @@
+	if len(nodeIPs) < 2 {
+		test.Logf("Not enough nodes with XFRM states for validation (have %d, need 2)", len(nodeIPs))
+		return
+	}
```
Lines 71–80 are not needed anymore if we fatal on line 63.

pillai-ashwin replied:
Removed that section

pchaigno commented on cilium-cli/connectivity/tests/ipsec_key_derivation.go:
```diff
@@ -0,0 +1,187 @@
+
+			for tunnelKey, state1 := range states1 {
+				if state2, exists := states2[tunnelKey]; exists {
+					if t.validateKeysWith(test, state1, state2, node1, node2, tunnelKey) {
```
If we find `state2` but its key doesn't match `state1`'s key, then we already know there's an issue and can fatal here directly. The advantage is that we can point the exact states with an issue in the fatal message.

pillai-ashwin replied:
* Removed the if condition around validateKeysWith()
* Changed validateKeysWith() to not return boolean, fails directly with specific state info

pchaigno commented on cilium-cli/connectivity/builder/ipsec_key_derivation.go:
```diff
@@ -0,0 +1,21 @@
+		WithCondition(func() bool {
+			ok, _ := ct.Features.MatchRequirements(features.RequireMode(features.EncryptionPod, "ipsec"))
+			return ok
+		}).
```
I think we should also only run this for `ct.Params().IncludeUnsafeTests` because of the potential scalability issue of the checks.

pillai-ashwin replied:
changed condition to return ok && ct.Params().IncludeUnsafeTests

pchaigno reviewed (changes requested):
A few suggestions for improvements below.

CI is also failing with what appears to be related (https://github.com/cilium/cilium/actions/runs/17728520295). It also looks like the test was skipped for configuration ipsec-7; I'm wondering why?

pchaigno commented on cilium-cli/connectivity/builder/ipsec_key_derivation.go:
```diff
@@ -0,0 +1,21 @@
+		WithCondition(func() bool {
+			ok, _ := ct.Features.MatchRequirements(features.RequireMode(features.EncryptionPod, "ipsec"))
+			return ok && ct.Params().IncludeUnsafeTests
+		}).
```
Why can't we have a Cilium version check here with `WithCiliumVersion`?

pillai-ashwin replied:
Ah, `WithCiliumVersion()` is the proper way to handle this. I've updated the implementation to use that.
